lib/descriptor/descriptor.js

```diff
@@ -0,0 +1,187 @@
+/*!
+ * descriptor.js - descriptor object for bcoin
+ * Copyright (c) 2023, the bcoin developers (MIT License).
+ * https://github.com/bcoin-org/bcoin
+ */
+
+'use strict';
+
+const assert = require('bsert');
+const common = require('./common');
+
+/**
+ * Constants
+ */
+
+/**
+ * String type for key in a descriptor.
+ * stringType.PUBLIC for descriptor with public key(s)
+ * stringType.PRIVATE for descriptor with private key(s)
+ * @const {Object}
+ */
+
+const stringType = {
+  PUBLIC: 'PUBLIC',
+  PRIVATE: 'PRIVATE'
+};
+
+/**
+ * Descriptor
+ * Base class for all descriptors.
+ * Represents an output script.
+ * @alias module:descriptor.Descriptor
+ * @see https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md
+ * @property {String} name - script expression of descriptor function.
+ * @property {KeyProvider[]} keyProviders - parsed key arguments
+ * for the descriptor
+ * (size 1 for PK, PKH, WPKH; any size for WSH and Multisig).
+ * @property {Descriptor[]} subdescriptors - sub-descriptor arguments
+ * for the descriptor (empty for everything but SH and WSH)
+ * @property {Network} network
+ */
+
+class Descriptor {
+  /**
+   * Create a descriptor
+   * @constructor
+   */
+
+  constructor() {
+    this.name = null;
+    this.keyProviders = [];
+    this.subdescriptors = [];
+    this.network = null;
+  }
```

Comment:
See my comment about making `AbstractKeyProvider` - I think more stuff can be moved to the base class, like parsing some of the common options like network. `keyProviders` must always be an array so that can be checked in base class, then each extended class can check those array lengths, etc.